### PR TITLE
[CL-76] Handle cat profile updates

### PR DESCRIPTION
**Pull request by guarde**

### CODE REVIEW

File: prisma/schema.prisma [hunk 1]
```diff
@@ -16,9 +16,11 @@ model User {
   username String @unique
 
   devices Device[]
+  cats    Cat[]
 
   updated_at DateTime @updatedAt
   created_at DateTime @default(now())
+  deleted_at DateTime?
```
Comment (stevelw) [inline prisma/schema.prisma]: Is this GDPR compliant? I don't think it is.
What's the reason for adding soft deletion?

File: prisma/schema.prisma [hunk 2]
```diff
@@ -38,4 +42,22 @@ model Device {
...
+}
+
+model Cat {
+  id String @unique @default(cuid())
+
+  name        String @default("Cat")
+  description String @default("Meow meow. Meow meow? Meow!")
```
Comment (stevelw) [inline prisma/schema.prisma]: Why are these set to defaults?

File: src/controllers/cats.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+		name: 'string,optional',
+		description: 'string,optional',
+		picture_url: 'string,optional',
+	};
+	const payload = request.body;
+	const result = validator(payload, schema);
+
+	if (!result.success) return next({ status: 400, message: result.errors });
+
+	cats.createCat(user.username, result.body).then((cat) => {
```
Comment (stevelw) [inline src/controllers/cats.ts]: Should this be user's id (generated by us) not username (generated by them) to guard against corrupting inputs?

File: src/controllers/cats.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	};
+
+	const payload = request.body;
+	const result = validator(payload, schema);
+
+	if (!result.success) return next({ status: 400, message: result.errors });
+
+	cats.fetchCatsByUserID(user.id).then((usersCats) => {
+		// ensure that the authenticated user owns the cat
+		const catsThatMatch = usersCats.filter(
```
Comment (stevelw) [inline src/controllers/cats.ts]: If the ID is unique, then find() would be a lot more efficient.

Comment (stevelw) [reply src/controllers/cats.ts]: Is usersCats an array? If so, you could just use .includes()

File: src/controllers/cats.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+			(cat) => cat.id === result.body.cat_id
+		);
+		if (catsThatMatch.length === 0)
+			return next({ status: 404, message: 'Cat not found' });
+		delete result.body.cat_id;
```
Comment (stevelw) [inline src/controllers/cats.ts:50]: I don't see the purpose of this line?

File: prisma/schema.prisma (hunk 1 above)
Comment (guarde) [reply prisma/schema.prisma]: Sorry - I stole this code from one of my previous repo's to fix an issue with relation deletions. The reason my previous repo does it is to establish and defend legal claims against explicit content - It isn't needed here so I will switch

File: prisma/schema.prisma (hunk 2 above)
Comment (guarde) [reply prisma/schema.prisma]: Initial thought process was to have something to display on frontend. I suppose only an image should have a default

File: src/controllers/cats.ts (hunk 3 above)
Comment (guarde) [reply src/controllers/cats.ts]: This was pre-middleware that I forgot to change fully. Will work on it

File: prisma/schema.prisma (hunk 2 above)
Comment (guarde) [reply prisma/schema.prisma]: On second thought, I guess we could just have a default on the client side.

File: src/controllers/cats.ts (hunk 3 above)
Comment (guarde) [reply src/controllers/cats.ts]: Although in theory there should never be a case where a username should be able to corrupt this - If that was true, we have a bigger issue on our hands. Still gonna change over to a new method tho

File: src/controllers/cats.ts (hunk 4 above)
Comment (guarde) [reply src/controllers/cats.ts]: .includes wont work as its an array of objects. find would be more efficient

**Review by stevelw — APPROVED** [review]

**Pull request merged**